add list_collections endpoint, solves #1835 (#1856)
huggingface/huggingface_hub: docs/source/en/guides/collections.md

In addition to these base attributes, returned items can have additional attributes depending on their type: `author`, `private`, `lastModified`, `gated`, `title`, `likes`, `upvotes`, etc. None of these attributes are guaranteed to be returned.

## List collections

> We can also get multiple collections with [`list_collection`]. It allows us to retrieve collections and filter them with certain parameters. Let's list all collections of the user `teknium`.

Reviewer: ```suggestion
We can also retrieve collections using [`list_collections`]. Collections can be filtered using some parameters. Let's list all the collections from the user [`teknium`](https://huggingface.co/teknium).
```

```py
>>> from huggingface_hub import list_collections

>>> collections = list_collections(owner="teknium")
```

> This returns an iterable of `Collection` objects. We can iterate over it to print number of upvotes for each collections.

Reviewer: ```suggestion

This returns an iterable of `Collection` objects. We can iterate over them to print, for example, the number of upvotes for each collection.

```

(adding some newlines to help the doc builder parse correctly the markdown file)

```py

> >>> for i in collections:
> ...   print("Number of upvotes:", i.upvotes)

Reviewer: ```suggestion
>>> for collection in collections:
...   print("Number of upvotes:", collection.upvotes)
```

Number of upvotes: 1
Number of upvotes: 5
```

> It is possible to do more advanced filtering. Let's get all model collections, sort by trending, and limit the count to 5.

Reviewer: ```suggestion
It is possible to do more advanced filtering. Let's get all collections containing the model [TheBloke/OpenHermes-2.5-Mistral-7B-GGUF](https://huggingface.co/TheBloke/OpenHermes-2.5-Mistral-7B-GGUF), sorted by trending, and limit the count to 5.
```

While testing the API I noticed that filtering by item allow to list collections containing a specific item, not only listing collections with "models". I think it's a good occasion to showcase a cool model added to a lot of collections (like [TheBloke/OpenHermes-2.5-Mistral-7B-GGUF](https://huggingface.co/TheBloke/OpenHermes-2.5-Mistral-7B-GGUF))

```py

> >>> collections = list_collections(item="models", sort="trending", limit=5):
> >>> for i in collections:
> ...   print("Number of upvotes:", i.upvotes)
> Number of upvotes: 39
> Number of upvotes: 178
> Number of upvotes: 21
> Number of upvotes: 11
> Number of upvotes: 14

Reviewer: ```suggestion
>>> collections = list_collections(item="models/TheBloke/OpenHermes-2.5-Mistral-7B-GGUF", sort="trending", limit=5):
>>> for collection in collections:
...   print(collection.slug)
teknium/quantized-models-6544690bb978e0b0f7328748
AmeerH/function-calling-65560a2565d7a6ef568527af
PostArchitekt/7bz-65479bb8c194936469697d8c
gnomealone/need-to-test-652007226c6ce4cdacf9c233
Crataco/favorite-7b-models-651944072b4fffcb41f8b568
```

(adapting to filter by specific item + showcase that we can also get the slug (to show something different than the previous example))

```

> `sort` accepts one of the following values:
> * `last_modified`
> * `trending`
> * `upvotes`
> 
> `item` accepts `spaces`, `models`, `datasets`, or `papers`. It is possible to specify a particular item, too. For example:
> * `models/teknium/OpenHermes-2.5-Mistral-7B`
> * `spaces/julien-c/open-gpt-rhyming-robot`
> * `datasets/squad`
> * `papers/2311.12983`

Reviewer: ```suggestion
Parameter `sort` must be one of  `"last_modified"`,  `"trending"` or `"upvotes"`. Parameter `item` accepts any particular item. For example:
* `"models/teknium/OpenHermes-2.5-Mistral-7B"`
* `"spaces/julien-c/open-gpt-rhyming-robot"`
* `"datasets/squad"`
* `"papers/2311.12983"`

For more details, please check out [`list_collections`] reference.
```

It looks like supporting generic search like `item="spaces"` is more a bug than a feature :confused: I will open an issue on the server repo to fix this. For example, if I go to https://huggingface.co/api/collections?sort=upvotes&item=spaces, the first collection listed is https://huggingface.co/collections/TheBloke/recent-models-last-100-repos-sorted-by-creation-date-64f9a55bb3115b4f513ec026 which contains only models... We will fix this but the safest is to document only the "search by specific item" use case.
